### PR TITLE
Remplace named tuples with enums

### LINKED ISSUE #1222: Use Enums instead of NamedTuple
## Overview

Currently we use `NamedTuple` for the static parameters to send to the API.
Replace with `Enum` instances, much better to deal with.

## Details

The named tuples are more complexe to type and to work with.

With the enums it's much easier and faster.

## Mention

Add warnings again in the latest release for the coming breaking change :grimacing:

### PR DESCRIPTION
**Pull request by lavigne958**
Most of the parameters for the API are `ENUM`s

Replace the named tuples that are not fitted for that with enums.

Each enum hold one of the choices.
The actual string value must be extracted in order to pass strings only to the API using the function `extract_enum_value` using the `enum.value` field if the enum is not `None`.

this forces us to filter all arguments that are passed when they can possibly be `None`.

thought about setting the default value for each instead but this changes the API call made to the API and that could lead to unwanted behavior for the users.

Possible way was to add a new value to each enum: `default = None` which will result in a `None` value so it won't be part of the API call, but this requires all method to have that value as default value, and if a user (for some reason) sets one of the parameter to `param=None` then the code breaks and requires that user to change to `param=MyEnum.none` which is not user friendly. To me any function with default empty param should be either `None` or some default container like `param=[]` etc.

so far this is the less intrusive for users and keeps the API calls as is and use the default behavior from google.

closes #1222

### CODE REVIEW

File: gspread/utils.py
Comment (alifeee) [inline gspread/utils.py:67]: should we consider including `formated_string` for backwards compatibility? probably not.

File: gspread/worksheet.py
Comment (alifeee) [inline gspread/worksheet.py:1093]: why not use extract_enum_values here?

**Review by alifeee — APPROVED** [review]

**Comment by alifeee** [comment]
I cannot find the comment we discussed it originally, but why not use strEnums, so that you do not have to use `.value`?

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py:1093]: True I could, I will

**Comment by lavigne958** [comment]
> I cannot find the comment we discussed it originally, but why not use strEnums, so that you do not have to use `.value`?

We can't use `StrEnum` they were introduced in python3.11 only :-( I though about it, it would make life much easier !

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py:1093]: I remember why I did not use the function here:
- the function `extract_enum_values` only takes dict as argument
- So I need to create a dict just to pass the single value
- then get the resulting dict
- if a value put it, if no value put `None`

I felt like this was easier to put

**Comment by lavigne958** [comment]
> > We can't use StrEnum they were introduced in python3.11 only :-( I though about it, it would make life much easier !
> 
> What about with StrEnum: https://pypi.org/project/StrEnum/
> 
> It supports Python 3.7+

wow awesome ! :star_struck:  I'll update the code to use it then it makes like so much easier (If everything goes according to plan lol)

File: gspread/worksheet.py
Comment (alifeee) [inline gspread/worksheet.py:1206]: these two `body` modifications can go in the object now, instead of afterwards (they were not before because of parse_enum function)

File: gspread/worksheet.py
Comment (alifeee) [inline gspread/worksheet.py:1701]: can this go back in the `params` object?

Comment (lavigne958) [reply gspread/worksheet.py:1701]: Yes or can. Good catch.

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py:1206]: Yes it can too. I forgot with  all the back and forth

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py]: done

File: gspread/worksheet.py
Comment (lavigne958) [reply gspread/worksheet.py]: done

**Comment by lavigne958** [comment]
you already approved but as I made changes I wait for your review before merging.

**Review by alifeee — APPROVED** [review]
lgtm!

it's nice that worksheet only has docs changes now that the enums can be used where strings are used.

question: why all the whitespace change?

image

ready to merge! good work! ❤

**Comment by lavigne958** [comment]
> lgtm!
> 
> it's nice that worksheet only has docs changes now that the enums can be used where strings are used.

Yes that's right, so much easier.

> 
> question: why all the whitespace change?

Because the RST documentation format works with indentation. The length (number of whitespaces) need to create a block, a list, a warning etc is aligned with the parent block. All of those where not aligned so I aligned them to properly render the documentation.
> 
> image
> 
> ready to merge! good work! ❤

**Pull request merged**
